### PR TITLE
Render plugins remodularization

### PR DESCRIPTION
**Pull request by AgustinBettati**

### CODE REVIEW

File: shared/src/main/scala/amf/client/remod/AMFEnvironment.scala [hunk 1]
```diff
@@ -29,7 +29,7 @@ abstract private[amf] class BaseEnvironment(val resolvers: AMFResolvers,
 
   def withPlugin(amfPlugin: AMFPlugin[_]): Self = doCopy(registry.withPlugin(amfPlugin))
 
-  def removePlugin(amfPlugin: AMFPlugin[_]): Self = doCopy(registry.removePlugin(amfPlugin))
+  def removePlugin(id: String): Self = doCopy(registry.removePlugin(id))
```
Comment (tomsfernandez) [inline shared/src/main/scala/amf/client/remod/AMFEnvironment.scala:32]: Great change

File: shared/src/main/scala/amf/client/remod/AMFEnvironment.scala [hunk 2]
```diff
@@ -56,6 +56,7 @@ private[amf] case class AMFEnvironment(override val resolvers: AMFResolvers,
   type Self = AMFEnvironment
 
   def withParsingOptions(parsingOptions: ParsingOptions): AMFEnvironment = this.copy(options = options.copy(parsingOptions = parsingOptions))
+  def withRenderingOptions(renderingOptions: RenderOptions): AMFEnvironment = this.copy(options = options.copy(renderingOptions = renderingOptions))
```
Comment (tomsfernandez) [inline shared/src/main/scala/amf/client/remod/AMFEnvironment.scala]: probably leave `withRenderOptions` to have "coherence" between the options object name and the setter.

File: shared/src/main/scala/amf/client/remod/amfcore/config/RenderOptions.scala [hunk 3]
```diff
@@ -2,11 +2,11 @@ package amf.client.remod.amfcore.config
 
 import amf.core.errorhandling.{ErrorHandler, UnhandledErrorHandler}
 import amf.core.metamodel.Field
-
+import amf.core.emitter.{RenderOptions => LegacyRenderOptions}
 /**
   * Immutable implementation of render options
   */
-private[remod] case class RenderOptions(
+private[amf] case class RenderOptions(
```
Comment (tomsfernandez) [inline shared/src/main/scala/amf/client/remod/amfcore/config/RenderOptions.scala:8]: Is the `eh` left for backwards compatibility only? If it is I'd add a big TODO to erase it afterwards.

File: shared/src/main/scala/amf/core/AMFSerializer.scala [hunk 4]
```diff
@@ -114,31 +112,36 @@ class AMFSerializer(unit: BaseUnit,
...
+            mediaType: String,
+            vendor: String,
+            options: RenderOptions,
+            shapeOptions: ShapeRenderOptions = ShapeRenderOptions()): AMFSerializer = {
+    val immutableOptions = RenderOptions.toImmutable(options)
+    val env = AMFPluginsRegistry.obtainStaticEnv().withRenderingOptions(immutableOptions)
+    new AMFSerializer(unit, mediaType, vendor, shapeOptions, env)
+  }
+
   def init()(implicit executionContext: ExecutionContext): Unit = {
```
Comment (tomsfernandez) [inline shared/src/main/scala/amf/core/AMFSerializer.scala:162]: Can't we get rid of this `init` method?

File: shared/src/main/scala/amf/client/remod/amfcore/plugins/render/AMFRenderPlugin.scala [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  def emit[T](unit: BaseUnit,
+              builder: DocBuilder[T],
+              renderOptions: RenderOptions = RenderOptions(),
+              shapeRenderOptions: ShapeRenderOptions = ShapeRenderOptions()): Boolean
+}
+
+
+private[amf] case class AMFRenderPluginAdapter(plugin: AMFDocumentPlugin) extends AMFRenderPlugin {
+
+  override def emit[T](unit: BaseUnit, builder: DocBuilder[T], renderOptions: RenderOptions, shapeRenderOptions: ShapeRenderOptions): Boolean =
```
Comment (tomsfernandez) [inline shared/src/main/scala/amf/client/remod/amfcore/plugins/render/AMFRenderPlugin.scala]: Isn't it weird that the `emit` method returns a boolean? I know it's for backwards compatibility but still. Weird.

File: shared/src/main/scala/amf/core/AMFSerializer.scala [hunk 6]
```diff
@@ -114,31 +112,36 @@ class AMFSerializer(unit: BaseUnit,
...
-    } match {
-      case Some(domainPlugin) =>
-        Some(domainPlugin)
-      case None => AMFPluginsRegistry.documentPluginForMediaType(mediaType).find(_.canUnparse(unit))
-    }
 
-  private def getDomainPlugin: AMFDocumentPlugin =
-    findDomainPlugin().getOrElse {
+  private def getRenderPlugin: AMFRenderPlugin = {
+    val renderPlugin = env.registry.plugins.renderPlugins.sorted.find(_.applies(RenderInfo(unit, vendor, mediaType)))
```
Comment (tomsfernandez) [inline shared/src/main/scala/amf/core/AMFSerializer.scala]: The fact that we have to call `sorted` all the time by hand means that the `renderPlugins` list should be a SortedList

File: shared/src/main/scala/amf/client/remod/amfcore/config/RenderOptions.scala (hunk 3 above)
Comment (AgustinBettati) [reply shared/src/main/scala/amf/client/remod/amfcore/config/RenderOptions.scala:8]: eh is deleted from the RenderOptions defined in remod package. Client and internal (legacy) render options still have the error handler.

File: shared/src/main/scala/amf/client/remod/AMFEnvironment.scala (hunk 2 above)
Comment (AgustinBettati) [reply shared/src/main/scala/amf/client/remod/AMFEnvironment.scala]: good catch

File: shared/src/main/scala/amf/core/AMFSerializer.scala (hunk 6 above)
Comment (AgustinBettati) [reply shared/src/main/scala/amf/core/AMFSerializer.scala]: This is a valid point, and also applies to parsing plugins for example.

File: shared/src/main/scala/amf/core/AMFSerializer.scala (hunk 4 above)
Comment (AgustinBettati) [reply shared/src/main/scala/amf/core/AMFSerializer.scala:162]: Would wait for when we delete RuntimeSerializer and this will become obsolete.

**Review by tomsfernandez — APPROVED** [review]

**Pull request merged**